File: 0042-trapping-rain-water/0042-trapping-rain-water.java
class Solution {
    public int trap(int[] height) {
        int l = 0 ;
        int n = height.length;
        int r = n - 1;
        int TrappedWater = 0;
        int Lmax = 0;
        int Rmax = 0;
        while(l<r){
            if(height[l] < height[r]){
                if(height[l] >= Lmax){
                    Lmax = height[l];
                }
                else{
                    TrappedWater += Lmax - height[l]; 
                }
                l++;
            }
            else{
                if(height[r] >= Rmax){
                    Rmax = height[r];
                }
                else{
                    TrappedWater += Rmax - height[r];
                }
                r--;
            }
        }
        return TrappedWater;

    }
}

File: 0042-trapping-rain-water/0042-trapping-rain-water.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        l=[0]*n
        r=[0]*n
       
        m1=0
        for i in range(n):
            m1=max(m1,height[i])
            l[i]=m1
        m2=0
        for i in range(n-1,-1,-1):
            m2=max(m2,height[i])
            print(m2)
            r[i]=m2
        trap=0
        for i in range(n):
            trap+=(min(l[i],r[i])-height[i])
        return trap